File: api/api-lambdas/putcontent/putcontent.js
const AWS = require('aws-sdk')
const { createMetricsLogger, Unit } = require("aws-embedded-metrics");

var personalizeevents = new AWS.PersonalizeEvents();
//var dynamoClient = new AWS.DynamoDB.DocumentClient();

console.log('Loading function');

function getTimeZoneOffset(date, timeZone) {

  // Abuse the Intl API to get a local ISO 8601 string for a given time zone.
  let iso = date.toLocaleString('en-CA', { timeZone, hour12: false }).replace(', ', 'T');

  // Include the milliseconds from the original timestamp
  iso += '.' + date.getMilliseconds().toString().padStart(3, '0');

  // Lie to the Date object constructor that it's a UTC time.
  const lie = new Date(iso + 'Z');

  // Return the difference in timestamps, as minutes
  // Positive values are West of GMT, opposite of ISO 8601
  // this matches the output of `Date.getTimeZoneOffset`
  return -(lie - date) / 60 / 1000;
}

var now = new Date();
var timezone_offset = (getTimeZoneOffset(now, process.env.EventsTZ) - 60) * 60 * 1000

exports.handler = (event, context, callback) => {
    console.log(JSON.stringify(event, null, 2));
    
    event.Records.forEach(function(record) {
        // Kinesis data is base64 encoded so decode here
        const metrics = createMetricsLogger();
        metrics.putDimensions({ Type: "PutContent" });

        var payload = Buffer.from(record.kinesis.data, 'base64').toString('ascii');
        console.debug('Decoded payload:', payload);
        payload = JSON.parse(payload);
        
        if (!payload.Published){
            console.debug('Skipping not published content', payload);
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }

        if (payload.Sponsored && payload.Sponsored === true){
            console.debug('Skipping Sponsored content', payload);
            metrics.putMetric("EventStatus", 0);
            metrics.flush();
            return context.successful;
        }
/**
        if ((payload.sp_user_id === undefined || payload.sp_user_id.trim().length === 0) && (payload.sp_domain_userid === undefined || payload.sp_domain_userid.trim().length === 0)){
            console.debug("Skipping event: missing sp_user_id and sp_domain_userid")
            return context.successful;
        }
    **/
        var eventDate = new Date(payload.UpdatedDate);
        var delay_ms = now.getTime() - eventDate.getTime() - timezone_offset;
        
    
        metrics.putMetric("DeliveryLatencyMS", delay_ms, Unit.Milliseconds);

        var putItemsParams= {
            'datasetArn': process.env.CONTENT_DATASET_ARN,
            'items': [
                {
                  'itemId': payload.ContentId,
                  'properties': {
                      'ContentText': payload.ContentText,
                      'Category': payload.Category,
                      'WordCount': payload.WordCount,
                      'Published': payload.Published,
                      'ContentType': payload.ContentType,
                      'CREATION_TIMESTAMP': Math.floor(eventDate.getTime() / 1000),
                  }
                },
            ]
        }

        console.log("THIS IS THE OBJECT = " + JSON.stringify(putItemsParams,null,3))
        console.log("THIS IS THE SOURCE OBJECT = " + JSON.stringify(payload,null,3))

        personalizeevents.putItems(putItemsParams, function (err, data) {
          if (err) {
            metrics.putMetric("EventStatus", -1);
            metrics.flush();
            console.log(err, err.stack); // an error occurred
          }
          else{ 
                metrics.putMetric("EventStatus", 1);
                metrics.flush();
                console.log("Success: " + JSON.stringify(data, null, 2)) 
                //console.log(data);           // successful response
               // putItemsParams['items'][0]['CREATION_TIMESTAMP']=putItemsParams['items'][0]['CREATION_TIMESTAMP'].getTime();
                const putEventsErrResponse = {
                    statusCode: 500,
                    body: JSON.stringify(err),
                };
                callback(null, putEventsErrResponse);
                const response = {
                    statusCode: 200,
                    body: JSON.stringify(putItemsParams),
                };
                callback(null, response);
          }
        });
    
    });
};

File: api/api-lambdas/putContent/putContent.py
import time
import json
import os
import base64

from boto3 import client

from botocore import config as boto_config
from botocore.exceptions import ClientError
from aws_embedded_metrics import metric_scope

config = boto_config.Config(
    connect_timeout=1, read_timeout=1,
    retries={'max_attempts': 2})

enviroment            = os.environ.get('Environment')
resources_prefix      = os.environ.get('ResourcesPrefix')
filter_app_id         = os.environ.get('FilterAppId')
personalize_cli       = client('personalize-events', config=config)
dynamo_client         = client('dynamodb', config=config)
settings_dynamo_table =  f"{resources_prefix}-{enviroment}-api-settings"


def _get_dynamo_settings(db, dynamo_table):
    try:
        personalize_list = []
        scan_kwargs = {
            'TableName': dynamo_table,
            'FilterExpression': "#status = :active ",
            'ExpressionAttributeNames': {
                "#status": "status",
            },
            'ExpressionAttributeValues': {
                ":active": {
                    'S': "active"
                }
            }
        }

        done = False
        start_key = None
        while not done:
            if start_key:
                scan_kwargs['ExclusiveStartKey'] = start_key
            response = db.scan(**scan_kwargs)
            personalize_list.extend(response.get('Items', []))
            start_key = response.get('LastEvaluatedKey', None)
            done = start_key is None

        return personalize_list

    except ClientError as e:
        print(f"Key Error: {e}")


@metric_scope
def handler(event, context, metrics):
    metrics.set_property("LambdaRequestId", context.aws_request_id)
    metrics.put_dimensions({"Type": "PutEvents"})
    settings = _get_dynamo_settings(dynamo_client, settings_dynamo_table)

    if len(settings) == 0:
        return {'statusCode': '500', 'body': json.dumps("No active trackers found in dynamo settings table")}

    status_code = "200"
    status_body = json.dumps("Success")

    success_events = 0
    fail_events = 0
    skip_events = 0

    for record in event['Records']:
        #Kinesis data is base64 encoded so decode here
        payload_str=base64.b64decode(record["kinesis"]["data"])
        payload = json.loads(payload_str)
        print("Decoded payload: " + str(payload))


        if (payload.get('Published') is not True):
            printf(f"Skipping content: invalid Published {payload.get('Published')}")
            skip_events += 1;
            continue

        exclude = 0

        if (payload.get('Sponsored') is True):
            exclude = 1
        elif (payload.get('Section') == "life/horoscopes"):
            exclude = 1
        elif ("zerocanada" in payload.get('Keywords', [])):
            exclude = 1
        elif ("omit" in payload.get('Keywords', [])):
            exclude = 1

        try:
            timestamp = int(round(time.mktime(time.strptime(payload.get('UpdatedDate')[:-5], '%Y-%m-%dT%H:%M:%S'))))
        except:
            timestamp = int(round(time.time()))

        putItemsParams = {
            'items': [
                {
                    'itemId': payload.get('ContentId'),
                    'properties': json.dumps({
                        'ContentText': payload.get('ContentText'),
                        'Category': payload.get('Category'),
                        'WordCount': payload.get('WordCount'),
                        'Published': payload.get('Published'),
                        'ContentType': payload.get('ContentType'),
                        'Exclude': exclude,
                        'CREATION_TIMESTAMP': timestamp,
                    })
                }
            ]
        }

        status_code = "200"
        status_body = json.dumps("Success")

        for tracker in settings:
            print(f"Put item to dataset {tracker.get('datasetArn').get('S')}")
            putItemsParams['datasetArn'] = tracker.get('datasetArn').get('S')
            print("This is the input object: " + str(putItemsParams))
            try:
                response = personalize_cli.put_items(**putItemsParams)
                print(f"put_items response: {response}")
            except ClientError as e:
                status_code = "500"
                status_body = f"Personalize Client Error: {e}"
                print(f"Personalize Client Error: {e}")
                fail_events += 1
            else:
                success_events += 1

    metrics.put_metric("SuccessItems", success_events, "None")
    metrics.put_metric("FailItems", fail_events, "None")
    metrics.put_metric("SkipItems", skip_events, "None")
    return {'statusCode': status_code, 'body': status_body}



